stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,419 @@
+    )
+
+    # normalization axes string and numeric indices
+    axes_norm = set(axes_net).intersection(set(axes_check_and_normalize(test_input_norm_axes, disallowed='S')))
+    axes_norm = "".join(a for a in axes_net if a in axes_norm)  # preserve order of axes_net
+    axes_norm_num = tuple(axes_net.index(a) for a in axes_norm)
+
+    # normalize input image
+    test_input_norm = normalize(test_input, pmin=min_percentile, pmax=max_percentile, axis=axes_norm_num)
+
+    net_axes_in = axes_net.lower()
+    net_axes_out = axes_check_and_normalize(model._axes_out).lower()
+    ndim_tensor = len(net_axes_out) + 1
+
+    input_min_shape = list(axes_net_div_by)
+    input_min_shape[axes_net.index('C')] = model.config.n_channel_in
+    input_step = list(axes_net_div_by)
+    input_step[axes_net.index('C')] = 0
+
+    # add the batch axis to shape and step
+    input_min_shape = [1] + input_min_shape
+    input_step = [0] + input_step
+
+    # the axes strings in bioimageio convention
+    input_axes = "b" + net_axes_in.lower()
+    output_axes = "b" + net_axes_out.lower()
+
+    if mode == "keras_hdf5":
+        output_names = ("prob", "dist") + (("class_prob",) if model._is_multiclass() else ())
+        output_n_channels = (1, model.config.n_rays,) + ((1,) if model._is_multiclass() else ())
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        output_scale = [1]+list(1/g for g in model.config.grid) + [0]
+        output_offset = [0]*(ndim_tensor)
+
+    elif mode == "tensorflow_saved_model_bundle":
+        if model._is_multiclass():
+            raise NotImplementedError("Tensorflow SavedModel not supported for multiclass models yet")
+        # regarding input/output names: https://github.com/CSBDeep/CSBDeep/blob/b0d2f5f344ebe65a9b4c3007f4567fe74268c813/csbdeep/utils/tf.py#L193-L194
+        input_names = ["input"]
+        output_names = ["output"]
+        output_n_channels = (1 + model.config.n_rays,)
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        # same shape as input except for the channel dimension
+        output_scale = [1]*(ndim_tensor)
+        output_scale[output_axes.index("c")] = 0
+        # no offset, except for the input axes, where it is output channel / 2
+        output_offset = [0.0]*(ndim_tensor)
+        output_offset[output_axes.index("c")] = output_n_channels[0] / 2.0
+
+    assert all(s in (0, 1) for s in output_scale), "halo computation assumption violated"
+    halo = model._axes_tile_overlap(output_axes.replace('b', 's'))
+    halo = [int(np.ceil(v/8)*8) for v in halo]  # optional: round up to be divisible by 8
+
+    # the output shape needs to be valid after cropping the halo, so we add the halo to the input min shape
+    input_min_shape = [ms + 2 * ha for ms, ha in zip(input_min_shape, halo)]
```

Comment:
> I guess I still don't understand what halo really means.

It specifies the pixels that should be cropped (from one side) during tiled prediction, so that for a tile the shape is:
`out_tile_shape = in_tile_shape - 2 * halo`. ilastik and deepImageJ were not happy with the fact that we have allowed for configurations where `min_shape - 2 * halo <= 0`, i.e. tiling cannot be performed with the min shape, because this lead to some actual issues. Hence, we have changed the spec validation to mark this as invalid.

> I mean, isn't a consequence of increasing input_min_shape that I can't apply the model to small images anymore?

We currently don't check if inputs are bigger than `min_shape` when running normal prediction with `bioimageio.core`, so prediction with smaller images would still work, if the model can run it. It will only lead to issues when trying to run prediction with tiling.
I think that we should keep this in mind and in the future raise a warning if the input shape is smaller than min-shape, but still pass the input to the model and see if it can run prediction.

Note that I just changed `input_min_shape` in the most straight forward way I could see so that `min_shape - 2 * halo > 0`. Go ahead and change this if you see a better solution for this. (But this condition needs to be fulfilled now in order to have a valid model).

Reply:
> Hence, we have changed the spec validation to mark this as invalid.

Ok.

> I think that we should keep this in mind and in the future raise a warning if the input shape is smaller than min-shape, but still pass the input to the model and see if it can run prediction.

Not ideal, but ok.

> Go ahead and change this if you see a better solution for this.

`input_min_shape` will now be enlarged if necessary (https://github.com/stardist/stardist/pull/171/commits/6ab6d5f1e88100a6c756282d568794085b04992d)

I also updated the `tags` (https://github.com/stardist/stardist/pull/171/commits/db86a7ad052ce810a4c91ada4eb1aa322a82f5f2) to get rid of all the warnings.